File: android/src/main/kotlin/com/tapnetic/flutter_ezviz/flutter_ezviz/EzvizManager.kt
package com.tapnetic.flutter_ezviz.flutter_ezviz

import com.videogo.exception.BaseException
import com.videogo.openapi.EZConstants
import com.videogo.openapi.EZHCNetDeviceSDK
import com.videogo.openapi.EZGlobalSDK
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

object EzvizManager {

    private val ptzKeys = mapOf(
        Action_START to EZConstants.EZPTZAction.EZPTZActionSTART.name,
        Action_STOP to EZConstants.EZPTZAction.EZPTZActionSTOP.name,
        Command_Left to EZConstants.EZPTZCommand.EZPTZCommandLeft.name,
        Command_Right to EZConstants.EZPTZCommand.EZPTZCommandRight.name,
        Command_Up to EZConstants.EZPTZCommand.EZPTZCommandUp.name,
        Command_Down to EZConstants.EZPTZCommand.EZPTZCommandDown.name,
        Command_ZoomIn to EZConstants.EZPTZCommand.EZPTZCommandZoomIn.name,
        Command_ZoomOut to EZConstants.EZPTZCommand.EZPTZCommandZoomOut.name
    )

    fun sdkVersion(result: Result) {
        result.success(EZGlobalSDK.getVersion())
    }

    fun initSDK(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            val appKey = it["appKey"] as? String ?: ""
            val accessToken = it["accessToken"] as? String ?: ""
            val enableLog = it["enableLog"] as? Boolean ?: false
            val enableP2P = it["enableP2P"] as? Boolean ?: false
            val application = ApplicationUtils.application
            application?.let { app ->
                val ret = EZGlobalSDK.initLib(app, appKey)
                EZGlobalSDK.enableP2P(enableP2P)
                EZGlobalSDK.showSDKLog(enableLog)
                EZGlobalSDK.getInstance().setAccessToken(accessToken)
                EZHCNetDeviceSDK.getInstance()
                result.success(ret)
            }
        }
    }

    fun enableLog(arguments: Any?) {
        val map = arguments as? Map<*, *>
        map?.let {
            val debug = it["enableLog"] as? Boolean ?: false
            EZGlobalSDK.showSDKLog(debug)
        }
    }

    fun enableP2P(arguments: Any?) {
        val map = arguments as? Map<*, *>
        map?.let {
            val debug = it["enableP2P"] as? Boolean ?: false
            EZGlobalSDK.enableP2P(debug)
        }
    }

    fun setAccessToken(arguments: Any?) {
        val map = arguments as? Map<*, *>
        map?.let {
            val accessToken = it["accessToken"] as? String ?: ""
            EZGlobalSDK.getInstance().setAccessToken(accessToken)
        }
    }

    fun getDeviceInfo(arguments: Any?, result: Result) {
        CoroutineScope(Dispatchers.IO).launch {
            val map = arguments as? Map<*, *>
            map?.let {
                try {
                    val deviceSerial = it["deviceSerial"] as? String ?: ""
                    val deviceInfo = EZGlobalSDK.getInstance().getDeviceInfo(deviceSerial)
                    withContext(Dispatchers.Main) {
                        if (deviceInfo == null) {
                            result.success(null)
                        } else {
                            val device = mapOf(
                                "deviceSerial" to deviceInfo.deviceSerial,
                                "deviceName" to deviceInfo.deviceName,
                                "isSupportPTZ" to deviceInfo.isSupportPTZ,
                                "cameraNum" to deviceInfo.cameraNum
                            )
                            result.success(device)
                        }
                    }
                } catch (e: BaseException) {
                    withContext(Dispatchers.Main) {
                        result.error("获取设备异常", e.localizedMessage, null)
                    }
                }
            }
        }
    }

    fun getDeviceList(result: Result) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val deviceInfoList = EZGlobalSDK.getInstance().getDeviceList(0, 100)
                withContext(Dispatchers.Main) {
                    if (deviceInfoList.isNullOrEmpty()) {
                        result.success(null)
                    } else {
                        val array = deviceInfoList.map { deviceInfo ->
                            mapOf(
                                "deviceSerial" to deviceInfo.deviceSerial,
                                "deviceName" to deviceInfo.deviceName,
                                "isSupportPTZ" to deviceInfo.isSupportPTZ,
                                "cameraNum" to deviceInfo.cameraNum
                            )
                        }
                        result.success(array)
                    }
                }
            } catch (e: BaseException) {
                withContext(Dispatchers.Main) {
                    result.error("获取设备异常", e.localizedMessage, null)
                }
            }
        }
    }

    fun setVideoLevel(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            val deviceSerial = it["deviceSerial"] as? String ?: ""
            val cameraId = it["cameraId"] as? Int ?: 0
            val videoLevel = it["videoLevel"] as? Int ?: 0

            CoroutineScope(Dispatchers.IO).launch {
                val ret = EZGlobalSDK.getInstance().setVideoLevel(deviceSerial, cameraId, videoLevel)
                withContext(Dispatchers.Main) {
                    result.success(ret)
                }
            }
        }
    }

    fun controlPTZ(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            try {
                val deviceSerial = it["deviceSerial"] as? String ?: ""
                val cameraId = it["cameraId"] as? Int ?: 0
                val command = it["command"] as? String ?: ""
                val action = it["action"] as? String ?: ""
                val speed = it["speed"] as? Int ?: 0

                CoroutineScope(Dispatchers.IO).launch {
                    try {
                        val ret = EZGlobalSDK.getInstance().controlPTZ(
                            deviceSerial,
                            cameraId,
                            EZConstants.EZPTZCommand.valueOf(ptzKeys[command] ?: ""),
                            EZConstants.EZPTZAction.valueOf(ptzKeys[action] ?: ""),
                            speed
                        )
                        withContext(Dispatchers.Main) {
                            result.success(ret)
                        }
                    } catch (e: BaseException) {
                        withContext(Dispatchers.Main) {
                            result.error("云台控制异常", e.localizedMessage, null)
                        }
                    }
                }
            } catch (e: BaseException) {
                result.error("云台控制异常", e.localizedMessage, null)
            }
        }
    }

    fun loginNetDevice(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            val userId = it["userId"] as? String ?: ""
            val pwd = it["pwd"] as? String ?: ""
            val ipAddr = it["ipAddr"] as? String ?: ""
            val port = it["port"] as? Int ?: 0

            CoroutineScope(Dispatchers.IO).launch {
                try {
                    val netDevice = EZHCNetDeviceSDK.getInstance().loginDeviceWithUerName(userId, pwd, ipAddr, port)
                    withContext(Dispatchers.Main) {
                        result.success(
                            netDevice?.let {
                                mapOf(
                                    "userId" to netDevice.loginId,
                                    "channelCount" to netDevice.byChanNum,
                                    "startChannelNo" to netDevice.byStartChan,
                                    "dStartChannelNo" to netDevice.byStartDChan,
                                    "dChannelCount" to netDevice.byIPChanNum,
                                    "byDVRType" to netDevice.byDVRType
                                )
                            }
                        )
                    }
                } catch (e: BaseException) {
                    withContext(Dispatchers.Main) {
                        result.error("登录网络设备异常", e.localizedMessage, null)
                    }
                }
            }
        }
    }

    fun logoutNetDevice(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            val userId = it["userId"] as? Int ?: 0

            CoroutineScope(Dispatchers.IO).launch {
                val ret = EZHCNetDeviceSDK.getInstance().logoutDeviceWithUserId(userId)
                withContext(Dispatchers.Main) {
                    result.success(ret)
                }
            }
        }
    }

    fun netControlPTZ(arguments: Any?, result: Result) {
        result.error("Android不支持此方法", null, null)
    }
}

File: ios/Classes/EzvizManager.swift
//
//  EzvizManager.swift
//  flutter_ezviz
//
//  Created by 江鴻 on 2019/8/24.
//
import Foundation
import EZOpenSDKFramework
import Flutter


class EzvizManager {
    /// 获取SDK版本号
    static func sdkVersion(result: @escaping FlutterResult) {
        result(EZGlobalSDK.getVersion())
    }
    
    /// 初始化SDK
    static func initSDK(_ arguments: Any?,result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let appKey = map["appKey"] as? String ?? AppKey
            let accessToken = map["accessToken"] as? String ?? AccessToken
            let enableLog = map["enableLog"] as? Bool ?? false
            let enableP2P = map["enableP2P"] as? Bool ?? false
            let ret = EZGlobalSDK.initLib(withAppKey: appKey)
            EZGlobalSDK.setAccessToken(accessToken)
            EZGlobalSDK.setDebugLogEnable(enableLog)
            EZGlobalSDK.enableP2P(enableP2P)
            EZHCNetDeviceSDK.initSDK()
            result(ret)
        }else {
            result(false)
        }
    }
    
    /// 是否开启日志
    static func enableLog(_ arguments: Any?) {
        if let map = arguments as? Dictionary<String, Any> {
            if let debug = map["enableLog"] as? Bool {
                EZGlobalSDK.setDebugLogEnable(debug)
            }
        }
    }
    
    /// 是否开启P2P
    static func enableP2P(_ arguments: Any?) {
        if let map = arguments as? Dictionary<String, Any> {
            if let enableP2P = map["enableP2P"] as? Bool {
                EZGlobalSDK.enableP2P(enableP2P)
            }
        }
    }
    
    /// 设置Token
    static func setAccessToken(_ arguments: Any?) {
        if let map = arguments as? Dictionary<String, Any> {
            if let accessToken = map["accessToken"] as? String {
                EZGlobalSDK.setAccessToken(accessToken)
            }
        }
    }

    static func getDeviceInfo(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? [String: Any],
           let deviceSerial = map["deviceSerial"] as? String {
            EZGlobalSDK.getDeviceInfo(deviceSerial) { device, error in
                if let error = error {
                    ezvizLog(msg: error.localizedDescription)
                    result(nil)
                } else {
                    result(device.toJSON())
                }
            }
        } else {
            result(nil)
        }
    }
    
    static func getDeviceInfoList(result: @escaping FlutterResult) {
        EZGlobalSDK.getDeviceList(0, pageSize: 100) { (devices, count, error) in
            if let error = error {
                ezvizLog(msg: error.localizedDescription)
                result(FlutterError(code: "GET_DEVICE_LIST_ERROR", message: error.localizedDescription, details: nil))
                return
            }

            guard let devList = devices as? [EZDeviceInfo], !devList.isEmpty else {
                ezvizLog(msg: "Device list is empty or cannot be cast to [EZDeviceInfo]")
                result([])
                return
            }

            let devArr = devList.map { $0.toJSON() }
            result(devArr)
        }
    }
    
    /// 设置视频通道分辨率
    static func setVideoLevel(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let deviceSerial = map["deviceSerial"] as? String ?? ""
            let cameraId = map["cameraId"] as? Int ?? 0
            let videoLevel = map["videoLevel"] as? Int ?? 0
            // Updated to use the available case name `LevelLow`
            let ezvizVideoLevel = EZVideoLevelType.init(rawValue: videoLevel) ?? EZVideoLevelType.levelLow

            EZGlobalSDK.setVideoLevel(deviceSerial, cameraNo: cameraId, videoLevel: ezvizVideoLevel) { (error) in
                if let err = error {
                    ezvizLog(msg: err.localizedDescription)
                    result(false)
                } else {
                    result(true)
                }
            }
        }
    }
    
    /// 云台控制
    static func controlPTZ(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let deviceSerial = map["deviceSerial"] as? String ?? ""
            let cameraId = map["cameraId"] as? Int ?? 0
            let command = map["command"] as? String ?? ""
            let action = map["action"] as? String ?? ""
            let speed = map["speed"] as? Int ?? 0
            
            if let cmd = PTZKeys[command] as? EZPTZCommand{
                if let act = PTZKeys[action] as? EZPTZAction {
                    EZGlobalSDK.controlPTZ(deviceSerial, cameraNo: cameraId, command: cmd, action: act, speed: speed) { (error) in
                        if let err = error {
                            ezvizLog(msg: err.localizedDescription)
                            result(false)
                        }else {
                            result(true)
                        }
                    }
                    
                }else {
                    ezvizLog(msg: "action字符串不合法")
                    result(false)
                }
            }else {
                ezvizLog(msg: "command字符串不合法")
                result(false)
            }
        }
    }
}


// MARK: - 网络设备相关操作
extension EzvizManager {
    
    /// 登录网络设备
    static func loginNetDevice(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let userId = map["userId"] as? String ?? ""
            let pwd = map["pwd"] as? String ?? ""
            let ipAddr = map["ipAddr"] as? String ?? ""
            let port = map["port"] as? Int ?? 0
            
            DispatchQueue.global().async {
                let deviceInfo = EZHCNetDeviceSDK.loginDevice(withUerName: userId, pwd: pwd, ipAddr: ipAddr, port: port)
                
                DispatchQueue.main.async {
                    if let dev = deviceInfo {
                        result(dev.toJSON())
                    }else {
                        ezvizLog(msg: "无法登陆设备")
                        result(nil)
                    }
                }
            }
        }
    }
    
    /// 登出网络设备
    static func logoutNetDevice(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let userId = map["userId"] as? Int ?? 0
            result(EZHCNetDeviceSDK.logoutDevice(withUserId: userId))
        }
    }
    
    /// 网络设备云台控制
    static func netControlPTZ(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let userId = map["userId"] as? Int ?? 0
            let channelNo = map["channelNo"] as? Int ?? 0
            let command = map["command"] as? String ?? ""
            let action = map["action"] as? String ?? ""
            
            if let cmd = netPTZKeys[command] as? EZPTZCommandType{
                if let act = netPTZKeys[action] as? EZPTZActionType {
                    DispatchQueue.global().async {
                        let ret = EZHCNetDeviceSDK.ptzControl(withUserId: userId, channelNo: channelNo, command: cmd, action: act)
                        DispatchQueue.main.async {
                            result(ret)
                        }
                    }
                    
                }else {
                    ezvizLog(msg: "action字符串不合法")
                    result(false)
                }
            }else {
                ezvizLog(msg: "command字符串不合法")
                result(false)
            }
        }
    }
    
}
